src/domain-functions.ts

```diff
@@ -33,43 +25,33 @@ const makeDomainFunction: MakeDomainFunction =
       const result = inputSchema.safeParse(input)
 
       if (result.success === false) {
-        return { success: false, errors: [], inputErrors: result.error.issues }
+        return {
+          success: false,
+          errors: [],
+          inputErrors: result.error.issues,
+          data: null,
+        }
       } else if (envResult.success === false) {
         return {
           success: false,
           errors: envResult.error.issues,
           inputErrors: [],
+          data: null,
         }
       }
       try {
         return {
           success: true,
           data: await handler(result.data, envResult.data),
+          errors: [],
+          inputErrors: [],
         }
       } catch (error) {
         const errors = [{ message: (error as Error).message }]
-        return { success: false, errors, inputErrors: [] }
+        return { success: false, errors, inputErrors: [], data: null }
```

Comment:
I think I rather have `data` only in the success case. Here it seems to add more code without real benefit on the consumer side (since it was already `undefined` before).

Reply:
What bothers me with the data null in the type is that the `.data` in the result will type check even when we know that is a useless value (at least in the case of the empty lists they are mapped over).
In the code you have used as motivation we still have the conditionals in the form of the safe navigation and coalesce operators:
```
    Submitted: {actionData.data?.name ?? ''}
```

I would rather have the typechecker tell me data does not exist until I narrow the type, so I have a branch where success is guaranteed. I could even argue that usually the code above would be something like:

```
    {actionData.success ? 'Submitted: ' + {actionData.data.name} : 'Failure to sumit'}
```

Reply:
Ok, I think I didn't come up with the best example and the need might emerge but I'll agree with you for now and let it emerge if it ever will.

I'll rollback the nullish data for now